modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+          return {
+            label: (show_template ? '{{' + v.value + '}}: ' : '') + v.label,
+            value: v.value
+          };
+        }
+      });
+    if (list.length) return {
+        label: value.label,
+        list: list
+      };
+  });
+};
+
+Twinkle.block.callback.change_preset = function twinkleblockCallbackChangePreset(e) {
+  var $form = $(e.target.form), key = $form.find('[name=category]').val();
+  if (!key) return;
+
+  $form.find('[name=sub_group]').val(key);
+  var preset = Twinkle.block.blockPresetsInfo[key];
+  Twinkle.block.callback.update_form(e, preset);
+  Twinkle.block.callback.change_subcategory(e);
+};
+
+Twinkle.block.callback.change_expiry = function twinkleblockCallbackChangeExpiry(e) {
+  var $expiry = $(e.target.form).find('[name=expiry]');
+  if ($(e.target).val() === 'custom') {
+    $expiry.parent().show();
+  } else {
+    $expiry.parent().hide();
+    $expiry.val($(e.target).val());
+  }
+  e.target.form.block_timer.value = $expiry.val();
+};
+
+Twinkle.block.callback.update_form = function twinkleblockCallbackUpdateForm(e, data) {
+  var $form = $(e.target.form), $expiry_preset = $form.find('[name=expiry_preset]');
+
+  $expiry_preset.val(data.expiry || 'custom');
+
+  $form.find('[name=expiry]').val(data.expiry);
+  if ($expiry_preset.val() === 'custom') {
+    $form.find('[name=expiry]').parent().show();
+  } else {
+    $form.find('[name=expiry]').parent().hide();
+  }
+
+  // boolean-flipped options, more at [[mw:API:Block]]
+  data.disabletalk = data.disabletalk !== undefined ? data.disabletalk : (data.allowusertalk !== undefined ? data.allowusertalk !== '' : false);
+  data.hardblock = data.hardblock !== undefined ? data.hardblock : (data.anononly !== undefined ? data.anononly !== '' : false);
+
+  $form.find('[name=field_block_options]').find(':checkbox').each(function(i, el) {
+    // don't override original options if useInitialOptions is set
+    if (data.useInitialOptions && data[el.name] === undefined) return;
+    var check = data[el.name] === '' || !!data[el.name];
+    $(el).prop('checked', check);
+  });
+  $form.find('[name=reason]').val(data.reason);
+};
+
+Twinkle.block.callback.change_subcategory = function twinkleblockcallbackChangeSubcategory(e) {
+  var form = e.target.form, prev_block_timer = Twinkle.warn.prev_block_timer;
+  form.main_group.value = 'block';
+  Twinkle.warn.callback.change_subcategory(e);
+  if (prev_block_timer) form.expiry.value = prev_block_timer;
+};
+
+Twinkle.block.callback.evaluate = function twinkleblockCallbackEvaluate(e) {
+  var $form = $(e.target),
+      toBlock = $form.find('[name=actiontype][value=block]').is(':checked'),
+      toWarn = $form.find('[name=actiontype][value=template]').is(':checked');
+
+  if (!toBlock && !toWarn) return alert("Please give Twinkle something to do!");
+
+  if (toBlock) {
+    var blockoptions = $form.find('[name=field_block_options]').serializeObject(),
```

Comment:
Explained above, the hope was to give the block option checkboxes the same names as the API parameters, call this function and just hand over the data. Obviously some transformations had to be made, namely the annoying boolean inversion with `anononly` and `allowusertalk`. I actually didn't realize it wasn't part of core jQuery. I could rewrite this to map out the values with `serializeArray`